File: Algorithms/src/3483. Unique 3-Digit Even Numbers/Code/solution.py
class Solution:
    def totalNumbers(self, digits: List[int]) -> int:
        # I store how many times each digit appears.
        freq = [0] * 10

        # I build the frequency table so duplicate digits are handled correctly.
        for digit in digits:
            freq[digit] += 1

        # I store the total number of distinct valid 3-digit even numbers.
        answer = 0

        # I choose the hundreds digit from 1 to 9 because zero cannot be the first digit.
        for first in range(1, 10):
            # I choose the tens digit from 0 to 9.
            for second in range(10):
                # I choose only even digits for the ones position.
                for third in range(0, 10, 2):
                    # I skip this combination if any required digit is unavailable.
                    if freq[first] == 0 or freq[second] == 0 or freq[third] == 0:
                        continue

                    # Three equal digits require three copies of that digit.
                    if first == second == third and freq[first] < 3:
                        continue

                    # Equal first and second digits require two copies.
                    if first == second and freq[first] < 2:
                        continue

                    # Equal first and third digits require two copies.
                    if first == third and freq[first] < 2:
                        continue

                    # Equal second and third digits require two copies.
                    if second == third and freq[second] < 2:
                        continue

                    # This combination forms one distinct valid number.
                    answer += 1

        # I return the final count.
        return answer

File: Algorithms/src/3483. Unique 3-Digit Even Numbers/Code/solution.go
func totalNumbers(digits []int) int {
	// I store how many times each digit appears.
	freq := make([]int, 10)

	// I build the frequency table to handle duplicate digits correctly.
	for _, digit := range digits {
		freq[digit]++
	}

	// I store the number of distinct valid 3-digit even numbers.
	answer := 0

	// I choose the hundreds digit from 1 to 9 because zero cannot be the first digit.
	for first := 1; first <= 9; first++ {
		// I choose the tens digit from 0 to 9.
		for second := 0; second <= 9; second++ {
			// I choose only even digits for the ones position.
			for third := 0; third <= 8; third += 2 {
				// I skip this combination if any required digit is unavailable.
				if freq[first] == 0 || freq[second] == 0 || freq[third] == 0 {
					continue
				}

				// Three equal digits require three copies of that digit.
				if first == second && second == third && freq[first] < 3 {
					continue
				}

				// Equal first and second digits require two copies.
				if first == second && freq[first] < 2 {
					continue
				}

				// Equal first and third digits require two copies.
				if first == third && freq[first] < 2 {
					continue
				}

				// Equal second and third digits require two copies.
				if second == third && freq[second] < 2 {
					continue
				}

				// This combination represents one distinct valid number.
				answer++
			}
		}
	}

	// I return the total number of valid numbers.
	return answer
}